src/Storages/StorageMerge.cpp

```diff
@@ -295,6 +300,12 @@ void StorageMerge::read(
     auto modified_context = Context::createCopy(local_context);
     modified_context->setSetting("optimize_move_to_prewhere", false);
 
+    if (query_info.prewhere_info && !tableSupportsPrewhere())
```

Comment:
Hi @antonio2368 , I understand that this is almost zero sense to support this, but just out of curiosity, why not add a `CAST` instead of forbidden such cases? Or "zero sense" is the only argument?

Reply:
Sorry, I'm not sure where I said "zero sense" so I'm a bit confused.
This seemed the most straightforward and fastest fix for the issue. 
If the underlying table doesn't support prewhere not sure why `Merge` containing that table would. Or am I missing something?

Reply:
Sorry, seems that I did not explain it.

Here the question was about when the types in the tables are different, i.e. Nullable and non-Nullable types.
For this case `CAST` may work, right?

P.S. "zero sense" - no sense, in other words usually `Merge` is used for the identical tables, so making it work when the types differs is not a common use-case.

Reply:
ah I see, I think that's more connected to the initial PR (https://github.com/ClickHouse/ClickHouse/pull/44716)
and the question is why we disable prewhere instead of the cast when nullable is the differing type?
my solution was really ad-hoc and I assume it's true for the mentioned PR also, so I'm not saying that you're wrong.
If the inner table supports prewhere you only need to change the check `tableSupportsPrewhere` so it doesn't return false if `nullable` is the reason.

Reply:
Hi @antonio2368. I've recently found one more issue with PREWHERE for Merge, and come up with this patch #46454, where I removed this check, since it became redundant.